Aiida generator

agoscinski opened the pull request:
This PR uses the work in MR [!300](https://earth.bsc.es/gitlab/es/autosubmit/-/merge_requests/300) as base to add an entry point to the CLI to generate from autosubmit workflows an AiiDA version. I take a bit different approach than MR [!300](https://earth.bsc.es/gitlab/es/autosubmit/-/merge_requests/300) in the sense that no AiiDA object is generated in autosubmit but a completely self-contained folder with all scripts and files together with a script that runs the AiiDA workflow. This has the advantage that the aiida workflow can be generated without having aiida installed.

kinow commented:
Thanks for updating the branch and moving to GH. I've enabled the GH Actions and the tests should complete in some minutes :+1:

agoscinski committed:
Add generator entrypoint in CLI for aiida workflow generation

This PR uses the work in MR
[!300](https://earth.bsc.es/gitlab/es/autosubmit/-/merge_requests/300)
as base to add an entry point to the CLI to generate from autosubmit
workflows an AiiDA version. I take a bit different approach than MR
[!300](https://earth.bsc.es/gitlab/es/autosubmit/-/merge_requests/300)
in the sense that no AiiDA object is generated in autosubmit but a
completely self-contained folder with all scripts and files together
with a script that runs the AiiDA workflow. This has the advantage that
the aiida workflow can be generated without having aiida installed.

agoscinski commented:
@kinow could you run workflow?

kinow commented:
Sorry the delay @agoscinski , we are in the middle of releasing 4.1.15 and finishing up some issues but I'm slowly catching up with issues. Will reply your email in a bit. GH Actions running!

VindeeR commented:
Good afternoon @agoscinski, I'm still doing some tests, but I made a review of the main points to fix.

If anything needs to be clear or confused just let me know, thanks for you patience

agoscinski committed:
Replace yaml with ruamel usage

agoscinski committed:
Fix linting errors

Only exception is the usage of job._tmp_path has in the aiida generator.
I think that needs actually a refactor in the job class since JobPackage
also needs to use it. I am open for other suggestions but actually
reusing the logic in the jobs class to create the tmp_path that would
fix the linting seems to me to be the less robust solution.

agoscinski committed:
Add typehints everywhere

agoscinski committed:
use parse_time from Job and not ParamikoPlatform

agoscinski committed:
Add docstring to public API

agoscinski committed:
Sanitize expid used for linklabel

agoscinski committed:
Adapting tests to direct usage of autosubmit.generate_workflow

agoscinski commented on test/unit/test_generator.py:
I am sorry but the examples you pointed me are not enough for me to put autosubmit in a state where my custom as_conf is completely loaded. I think the tests you pointed to only work because they do not rely on a loaded job list unlike the test for the aiida generator. I spent over 2 hours to understand the creation process but the process is too convoluted for me to understand. I still create the job list the same way as before but now test the `autosubmit.generate_workflow` directly. The conf files should now be loaded from disk during the test. I hope that is okay.

agoscinski commented on autosubmit/generators/aiida.py:
done

agoscinski commented:
Thanks for running the CI, will fix type checking and formatting today

agoscinski committed:
Fix ruff errors

agoscinski commented:
Okay fixed ruff errors. The mypy errors are not related to the changes of this PR. I checked the lines that error out and none of them is touched by this PR

VindeeR reviewed (approved):
The changes looks fine, thanks for the hardwork 🙌 
Could you do a rebase before we merge it?

kinow commented on autosubmit/generators/__init__.py:
Missing license header.

kinow reviewed (commented):
Hi @agoscinski, I did a quick review this morning before the Autosubmit internal meeting. I left some comments, but here's a diff in case that helps more:

```diff
diff --git a/autosubmit/autosubmit.py b/autosubmit/autosubmit.py
index 9c4490ab..5f25ee13 100644
--- a/autosubmit/autosubmit.py
+++ b/autosubmit/autosubmit.py
@@ -38,7 +38,7 @@ from importlib.metadata import version
 from importlib.resources import files as read_files
 from pathlib import Path
 from time import sleep
-from typing import Generator, Optional, Union 
+from typing import Generator, Optional, Union
 
 from bscearth.utils.date import date2str
 from portalocker import Lock
@@ -58,6 +58,7 @@ from autosubmit.database.db_common import (
 from autosubmit.database.db_structure import get_structure
 from autosubmit.experiment.detail_updater import ExperimentDetails
 from autosubmit.experiment.experiment_common import copy_experiment, new_experiment, create_required_folders
+from autosubmit.generators import Engine, get_engine_generator
 from autosubmit.git.autosubmit_git import AutosubmitGit
 from autosubmit.git.autosubmit_git import check_unpushed_changes, clean_git
 from autosubmit.helpers.processes import process_id
@@ -79,7 +80,6 @@ from autosubmit.notifications.notifier import Notifier
 from autosubmit.platforms.paramiko_platform import ParamikoPlatform
 from autosubmit.platforms.paramiko_submitter import ParamikoSubmitter
 from autosubmit.platforms.platform import Platform
-from autosubmit.generators import Engine, get_engine_generator
 
 dialog = None
 
@@ -738,7 +738,8 @@ class Autosubmit:
                 'generate', description='Generate a workflow definition for a different workflow engine',
                 argument_default=argparse.SUPPRESS)
             subparser.add_argument('expid', help='experiment identifier')
-            subsubparser = subparser.add_subparsers(title="engines", dest='engine', required=True, description='Workflow engine identifier')
+            subsubparser = subparser.add_subparsers(
+                title="engines", dest='engine', required=True, description='Workflow engine identifier')
             for engine in Engine:
                 generator_class = get_engine_generator(engine)
                 parser_engine = subsubparser.add_parser(engine.value, help=f"{generator_class.get_engine_name()}")
diff --git a/autosubmit/generators/__init__.py b/autosubmit/generators/__init__.py
index d7096fab..101b64a7 100644
--- a/autosubmit/generators/__init__.py
+++ b/autosubmit/generators/__init__.py
@@ -1,13 +1,30 @@
+# Copyright 2015-2026 Earth Sciences Department, BSC-CNS
+#
+# This file is part of Autosubmit.
+#
+# Autosubmit is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Autosubmit is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Autosubmit.  If not, see <http://www.gnu.org/licenses/>.
+
 """This module provides generators to produce workflow configurations for different backend engines."""
+
+from abc import ABC, abstractmethod
 from enum import Enum
 from importlib import import_module
 from typing import TYPE_CHECKING
 
-from abc import ABC, abstractmethod
-
 if TYPE_CHECKING:
     import argparse
-    from autosubmitconfigparser.config.configcommon import AutosubmitConfig
+    from autosubmit.config.configcommon import AutosubmitConfig
     from autosubmit.job.job_list import JobList
 
 
@@ -79,7 +96,9 @@ def get_engine_generator(engine: Engine) -> AbstractGenerator:
     """
     return import_module(f'autosubmit.generators.{engine.value}').Generator
 
+
 __all__ = [
+    'AbstractGenerator',
     'Engine',
     'get_engine_generator'
 ]
diff --git a/autosubmit/generators/aiida.py b/autosubmit/generators/aiida.py
index feb283be..2d34049b 100644
--- a/autosubmit/generators/aiida.py
+++ b/autosubmit/generators/aiida.py
@@ -1,19 +1,37 @@
+# Copyright 2015-2026 Earth Sciences Department, BSC-CNS
+#
+# This file is part of Autosubmit.
+#
+# Autosubmit is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Autosubmit is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Autosubmit.  If not, see <http://www.gnu.org/licenses/>.
+
 """The AiiDA generator for Autosubmit."""
+
 import os
-from pathlib import Path
-from functools import cached_property
-import warnings
 import re
-from ruamel.yaml import YAML
+import warnings
+from functools import cached_property
+from pathlib import Path
 from typing import TYPE_CHECKING, Any
 
-from autosubmit.config.configcommon import AutosubmitConfig
+from ruamel.yaml import YAML
 
+from autosubmit.config.configcommon import AutosubmitConfig
+from autosubmit.generators import AbstractGenerator
 from autosubmit.job.job import Job
 from autosubmit.job.job_list import JobList
-from autosubmit.generators import AbstractGenerator
-from autosubmit.platforms.platform import Platform
 from autosubmit.platforms.locplatform import LocalPlatform
+from autosubmit.platforms.platform import Platform
 from autosubmit.platforms.slurmplatform import SlurmPlatform
 
 if TYPE_CHECKING:
@@ -56,7 +74,7 @@ class Generator(AbstractGenerator):
                               "DEPENDENCIES", "FILE", "RUNNING"]  # these are resolved by autosubmit internally
 
     SUPPORTED_PLATFORM_KEYWORDS = ["TYPE", "HOST", "USER", "QUEUE", "SCRATCH_DIR", "MAX_WALLCLOCK",
-                                    "PROJECT"]   # these are resolved by autosubmit internally
+                                   "PROJECT"]   # these are resolved by autosubmit internally
 
     def __init__(self, job_list: JobList, as_conf: AutosubmitConfig, output_dir: str):
         """
@@ -90,7 +108,8 @@ class Generator(AbstractGenerator):
         """
         output_dir = arg_options.get('output_dir', None)
         if not isinstance(output_dir, str):
-            raise ValueError(f"aiida generator requires output dir of type str but got {output_dir}") # TODO improve error message
+            # TODO improve error message
+            raise ValueError(f"aiida generator requires output dir of type str but got {output_dir}")
         self = cls(job_list, as_conf, output_dir)
         self._validate()
         workflow_script = self._generate_workflow_script()
@@ -218,7 +237,7 @@ tasks = {}
                     "label": f"{platform.name}",
                     "hostname": f"{platform.host}",
                     "work_dir": f"{platform.scratch}",
-                    #username": f"{platform.user}", does not work
+                    # username": f"{platform.user}", does not work
                     "description": "",
                     "transport": "core.ssh",
                     "scheduler": "core.slurm",
@@ -296,7 +315,9 @@ except NotExistent:
 
         for job in self._job_list.get_all():
             script_name = job.create_script(self._as_conf)
-            script_path = Path(job._tmp_path, script_name) # pylint: disable=protected-access
+            # noinspection PyProtectedMember
+            # pylint: disable=protected-access
+            script_path = Path(job._tmp_path, script_name)
             with open(script_path, "r", encoding="utf-8") as f:
                 script_text = f.read()
             # Let's drop the Autosubmit header and tailer.
@@ -370,15 +391,15 @@ example through a previous run, they are loaded. If you have changed the
 computer parameters in your autosubmit config, then you need to delete the
 corresponding compute and code nodes. Note that this has the disadvantage that
 it will delete all calculations that are associated with these nodes.
-It is therefore recommended to use the JOB paramaters in autosubmit to override
+It is therefore recommended to use the JOB parameters in autosubmit to override
 certain computer configs.
 """
+
     @staticmethod
     def _sanitize_label(label: str) -> str:
         return label.replace("-", "_")
 
 
-
 __all__ = [
     'Generator'
 ]
diff --git a/test/integration/generators/test_aiida.py b/test/integration/generators/test_aiida.py
new file mode 100644
index 00000000..79d4e8ff
--- /dev/null
+++ b/test/integration/generators/test_aiida.py
@@ -0,0 +1,124 @@
+# Copyright 2015-2026 Earth Sciences Department, BSC-CNS
+#
+# This file is part of Autosubmit.
+#
+# Autosubmit is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Autosubmit is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Autosubmit.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Integration tests for the AiiDA generator."""
+
+import argparse
+import os
+from typing import TYPE_CHECKING
+
+import pytest
+
+from autosubmit.generators import Engine
+
+if TYPE_CHECKING:
+    # noinspection PyProtectedMember
+    from py._path.local import LocalPath  # type: ignore
+
+
+@pytest.fixture()
+def jobs_data():
+    """Fixture with the jobs used for the AiiDA test experiments."""
+    return {
+        "JOBS": {
+            "LOCAL_SETUP": {
+                "FILE": "LOCAL_SETUP.sh",
+                "PLATFORM": "LOCAL",
+                "RUNNING": "once",
+                "DEPENDENCIES": {},
+                "ADDITIONAL_FILES": [],
+            },
+            "REMOTE_SETUP": {
+                "FILE": "REMOTE_SETUP.sh",
+                "PLATFORM": "LOCAL",
+                "DEPENDENCIES": {"LOCAL_SETUP": {}},
+                "WALLCLOCK": "00:05",
+                "RUNNING": "once",
+                "ADDITIONAL_FILES": [],
+            },
+            "INI": {
+                "FILE": "INI.sh",
+                "PLATFORM": "LOCAL",
+                "DEPENDENCIES": {"REMOTE_SETUP": {}},
+                "RUNNING": "member",
+                "WALLCLOCK": "00:05",
+                "ADDITIONAL_FILES": [],
+            },
+            "SIM": {
+                "FILE": "SIM.sh",
+                "PLATFORM": "LOCAL",
+                "DEPENDENCIES": {"INI": {}, "SIM-1": {}},
+                "RUNNING": "chunk",
+                "WALLCLOCK": "00:05",
+                "ADDITIONAL_FILES": [],
+            },
+            "POST": {
+                "FILE": "POST.sh",
+                "PLATFORM": "LOCAL",
+                "DEPENDENCIES": {"SIM": {}},
+                "RUNNING": "once",
+                "WALLCLOCK": "00:05",
+                "ADDITIONAL_FILES": [],
+            },
+            "CLEAN": {
+                "FILE": "CLEAN.sh",
+                "PLATFORM": "LOCAL",
+                "DEPENDENCIES": {"POST": {}},
+                "RUNNING": "once",
+                "WALLCLOCK": "00:05",
+                "ADDITIONAL_FILES": [],
+            },
+            "TRANSFER": {
+                "FILE": "TRANSFER.sh",
+                "PLATFORM": "LOCAL",
+                "DEPENDENCIES": {"CLEAN": {}},
+                "RUNNING": "member",
+                "ADDITIONAL_FILES": [],
+            },
+        }
+    }
+
+
+def test_aiida_generator2(
+        jobs_data,
+        autosubmit_exp: '',
+        tmp_path: 'LocalPath'):
+    exp = autosubmit_exp(experiment_data=jobs_data)
+    as_conf = exp.as_conf
+
+    # Test generator
+    generate_folder = tmp_path / "generate"
+    generate_folder.mkdir()
+    args = argparse.Namespace(output_dir=str(generate_folder))
+    exp.autosubmit.generate_workflow(exp.expid, Engine('aiida'), args)
+
+    expected_generated_paths_in_experiment_folder = {
+        f'{as_conf.expid}_LOCAL_SETUP.cmd', f'{as_conf.expid}_REMOTE_SETUP.cmd',
+        f'{as_conf.expid}_20000101_fc0_INI.cmd',
+        f'{as_conf.expid}_20000101_fc0_TRANSFER.cmd',
+        f'{as_conf.expid}_20000101_fc0_1_SIM.cmd',
+        f'{as_conf.expid}_20000101_fc0_2_SIM.cmd', f'{as_conf.expid}_POST.cmd',
+        f'{as_conf.expid}_CLEAN.cmd', 'local', "README.md",
+        "submit_aiida_workflow.py"
+    }
+    assert set(os.listdir(generate_folder / as_conf.expid)) == expected_generated_paths_in_experiment_folder
+
+    expected_generated_paths_in_local_folder = {
+        "bash@local-setup.yml",
+        "local-setup.yml"
+    }
+    assert set(os.listdir(generate_folder / as_conf.expid / "local")) == expected_generated_paths_in_local_folder
diff --git a/test/unit/test_generator.py b/test/unit/test_generator.py
deleted file mode 100644
index 8be5ba88..00000000
--- a/test/unit/test_generator.py
+++ /dev/null
@@ -1,231 +0,0 @@
-import pytest
-import os
-from ruamel.yaml import YAML
-
-@pytest.fixture()
-def jobs_data():
-    return {"JOBS": {
-            "LOCAL_SETUP": {
-                "FILE": "LOCAL_SETUP.sh",
-                "PLATFORM": "LOCAL",
-                "RUNNING": "once",
-                "DEPENDENCIES": {},
-                "ADDITIONAL_FILES": [],
-            },
-            "REMOTE_SETUP": {
-                "FILE": "REMOTE_SETUP.sh",
-                "PLATFORM": "LOCAL",
-                "DEPENDENCIES": {"LOCAL_SETUP": {}},
-                "WALLCLOCK": "00:05",
-                "RUNNING": "once",
-                "ADDITIONAL_FILES": [],
-            },
-            "INI": {
-                "FILE": "INI.sh",
-                "PLATFORM": "LOCAL",
-                "DEPENDENCIES": {"REMOTE_SETUP": {}},
-                "RUNNING": "member",
-                "WALLCLOCK": "00:05",
-                "ADDITIONAL_FILES": [],
-            },
-            "SIM": {
-                "FILE": "SIM.sh",
-                "PLATFORM": "LOCAL",
-                "DEPENDENCIES": {"INI": {}, "SIM-1": {}},
-                "RUNNING": "chunk",
-                "WALLCLOCK": "00:05",
-                "ADDITIONAL_FILES": [],
-            },
-            "POST": {
-                "FILE": "POST.sh",
-                "PLATFORM": "LOCAL",
-                "DEPENDENCIES": {"SIM": {}},
-                "RUNNING": "once",
-                "WALLCLOCK": "00:05",
-                "ADDITIONAL_FILES": [],
-            },
-            "CLEAN": {
-                "FILE": "CLEAN.sh",
-                "PLATFORM": "LOCAL",
-                "DEPENDENCIES": {"POST": {}},
-                "RUNNING": "once",
-                "WALLCLOCK": "00:05",
-                "ADDITIONAL_FILES": [],
-            },
-            "TRANSFER": {
-                "FILE": "TRANSFER.sh",
-                "PLATFORM": "LOCAL",
-                "DEPENDENCIES": {"CLEAN": {}},
-                "RUNNING": "member",
-                "ADDITIONAL_FILES": [],
-            },
-        }}
-
-
-@pytest.fixture()
-def expdef_data():
-    expid = "dummy-id"
-    return {
-        "DEFAULT": {"EXPID": f"{expid}", "HPCARCH": "LOCAL"},
-        "EXPERIMENT": {
-            "DATELIST": "20000101",
-            "MEMBERS": "fc0",
-            "CHUNKSIZEUNIT": "month",
-            "CHUNKSIZE": 4,
-            "NUMCHUNKS": 2,
-            "CHUNKINI": "",
-            "CALENDAR": "standard",
-        },
-        "PROJECT": {"PROJECT_TYPE": "none", "PROJECT_DESTINATION": ""},
-        "GIT": {
-            "PROJECT_ORIGIN": "",
-            "PROJECT_BRANCH": "",
-            "PROJECT_COMMIT": "",
-            "PROJECT_SUBMODULES": "",
-            "FETCH_SINGLE_BRANCH": True,
-        },
-        "SVN": {"PROJECT_URL": "", "PROJECT_REVISION": ""},
-        "LOCAL": {"PROJECT_PATH": ""},
-        "PROJECT_FILES": {
-            "FILE_PROJECT_CONF": "",
-            "FILE_JOBS_CONF": "",
-            "JOB_SCRIPTS_TYPE": "",
-        },
-        "RERUN": {"RERUN": False, "RERUN_JOBLIST": ""}
-    }
-
-@pytest.fixture()
-def platforms_data():
-    return {"PLATFORMS": {
-            "MARENOSTRUM4": {
-                "TYPE": "slurm",
-                "HOST": "mn1.bsc.es",
-                "PROJECT": "bsc32",
-                "USER": None,
-                "QUEUE": "debug",
-                "SCRATCH_DIR": "/gpfs/scratch",
-                "ADD_PROJECT_TO_HOST": False,
-                "MAX_WALLCLOCK": "48:00",
-                "TEMP_DIR": "",
-            },
-            "MARENOSTRUM_ARCHIVE": {
-                "TYPE": "ps",
-                "HOST": "dt02.bsc.es",
-                "PROJECT": "bsc32",
-                "USER": None,
-                "SCRATCH_DIR": "/gpfs/scratch",
-                "ADD_PROJECT_TO_HOST": False,
-                "TEST_SUITE": False,
-            },
-            "TRANSFER_NODE": {
-                "TYPE": "ps",
-                "HOST": "dt01.bsc.es",
-                "PROJECT": "bsc32",
-                "USER": None,
-                "ADD_PROJECT_TO_HOST": False,
-                "SCRATCH_DIR": "/gpfs/scratch",
-            },
-            "TRANSFER_NODE_BSCEARTH000": {
-                "TYPE": "ps",
-                "HOST": "bscearth000",
-                "USER": None,
-                "PROJECT": "Earth",
-                "ADD_PROJECT_TO_HOST": False,
-                "QUEUE": "serial",
-                "SCRATCH_DIR": "/esarchive/scratch",
-            },
-            "BSCEARTH000": {
-                "TYPE": "ps",
-                "HOST": "bscearth000",
-                "USER": None,
-                "PROJECT": "Earth",
-                "ADD_PROJECT_TO_HOST": False,
-                "QUEUE": "serial",
-                "SCRATCH_DIR": "/esarchive/scratch",
-            },
-            "NORD3": {
-                "TYPE": "SLURM",
-                "HOST": "nord1.bsc.es",
-                "PROJECT": "bsc32",
-                "USER": None,
-                "QUEUE": "debug",
-                "SCRATCH_DIR": "/gpfs/scratch",
-                "MAX_WALLCLOCK": "48:00",
-            },
-            "ECMWF-XC40": {
-                "TYPE": "ecaccess",
-                "VERSION": "pbs",
-                "HOST": "cca",
-                "USER": None,
-                "PROJECT": "spesiccf",
-                "ADD_PROJECT_TO_HOST": False,
-                "SCRATCH_DIR": "/scratch/ms",
-                "QUEUE": "np",
-                "SERIAL_QUEUE": "ns",
-                "MAX_WALLCLOCK": "48:00",
-            },
-        }}
-
-def test_aiida_generator2(mocker, autosubmit_config, autosubmit, expdef_data, jobs_data, platforms_data, tmp_path):
-    from autosubmit.config.basicconfig import BasicConfig
-    from autosubmit.generators import Engine 
-
-    expid = expdef_data['DEFAULT']['EXPID']
-    as_conf = autosubmit_config(expid, experiment_data=expdef_data)
-    mocker.patch('autosubmit.autosubmit.BasicConfig', as_conf.basic_config)
-    read_files_mock = mocker.patch('autosubmit.autosubmit.read_files', return_value=None)
-
-    config_path = tmp_path / expid / 'conf'
-    config_path.mkdir(parents=True, exist_ok=True)
-    BasicConfig.LOCAL_ROOT_DIR = tmp_path
-    yaml = YAML()
-    jobs_file_path = config_path / f"jobs_{expid}.yaml"
-    expdef_file_path = config_path / f"expdef_{expid}.yaml"
-    platforms_file_path = config_path / f"platforms_{expid}.yaml"
-
-    with open(jobs_file_path, "w") as f:
-        yaml.dump(jobs_data, f)
-    with open(expdef_file_path, "w") as f:
-        yaml.dump(expdef_data, f)
-    with open(platforms_file_path, "w") as f:
-        yaml.dump(platforms_data, f)
-    read_files_mock.return_value = tmp_path
-
-    # Generate job list data
-    from autosubmit.job.job_list import JobList
-    orig_generate = JobList.generate
-    def mockgenerate(self, *args, **kwargs):
-        kwargs['create'] = True
-        orig_generate(self, *args, **kwargs)
-    mocker.patch('autosubmit.job.job_list.JobList.generate', mockgenerate)
-    autosubmit.load_job_list(
-        as_conf.expid, as_conf, monitor=False
-    )
-
-    # Test generator
-    import argparse
-    generate_folder = tmp_path / "generate"
-    generate_folder.mkdir()
-    args = argparse.Namespace(output_dir=str(generate_folder.absolute()))
-    autosubmit.generate_workflow(expid, Engine('aiida'), args)
-
-    generated_paths_in_experiment_folder = set([
-        f'{as_conf.expid}_LOCAL_SETUP.cmd',
-        f'{as_conf.expid}_REMOTE_SETUP.cmd', # date from experiment data
-        f'{as_conf.expid}_20000101_fc0_INI.cmd',
-        f'{as_conf.expid}_20000101_fc0_TRANSFER.cmd', # date from experiment data
-        f'{as_conf.expid}_20000101_fc0_1_SIM.cmd', # date from experiment data
-        f'{as_conf.expid}_20000101_fc0_2_SIM.cmd', # date from experiment data
-        f'{as_conf.expid}_POST.cmd',
-        f'{as_conf.expid}_CLEAN.cmd',
-        'local',
-        "README.md",
-        "submit_aiida_workflow.py",
-    ])
-    assert set(os.listdir(generate_folder / as_conf.expid)) == generated_paths_in_experiment_folder
-
-    generated_paths_in_local_folder = set([
-        "bash@local-setup.yml",
-        "local-setup.yml"
-    ])
-    assert set(os.listdir(generate_folder / as_conf.expid / "local")) == generated_paths_in_local_folder
```

I rewrote the unit test as an integration test. The `autosubmit_config` is good only when you need to work with a "slim" experiment, and the AS config object. When you need other parts of the system, run commands, etc., it's easier to use an integration test. The new test didn't need any mocks, and I think it's doing the same. We still need to write better developer docs wiht info for new/external devs on how to write unit/integration/regression tests, sorry.

Running the new test locally works. There are several warnings in the output though, apparently from the code added in this PR (from the `_validate` function in `aiida.py`). Any idea if that's expected/normal, or if there is any way to fix those?

```bash
test/integration/generators/test_aiida.py::test_aiida_generator2
  /home/kinow/Development/python/workspace/autosubmit/autosubmit/generators/aiida.py:146: UserWarning: Found in job LOCAL_SETUP configuration file key ADDITIONAL_FILES that is not officially supported for AiiDA generator. It might result in an error.
    warnings.warn(msg)

test/integration/generators/test_aiida.py::test_aiida_generator2
  /home/kinow/Development/python/workspace/autosubmit/autosubmit/generators/aiida.py:146: UserWarning: Found in job REMOTE_SETUP configuration file key ADDITIONAL_FILES that is not officially supported for AiiDA generator. It might result in an error.
    warnings.warn(msg)

test/integration/generators/test_aiida.py::test_aiida_generator2
  /home/kinow/Development/python/workspace/autosubmit/autosubmit/generators/aiida.py:146: UserWarning: Found in job INI configuration file key ADDITIONAL_FILES that is not officially supported for AiiDA generator. It might result in an error.
    warnings.warn(msg)

test/integration/generators/test_aiida.py::test_aiida_generator2
  /home/kinow/Development/python/workspace/autosubmit/autosubmit/generators/aiida.py:146: UserWarning: Found in job SIM configuration file key ADDITIONAL_FILES that is not officially supported for AiiDA generator. It might result in an error.
    warnings.warn(msg)

test/integration/generators/test_aiida.py::test_aiida_generator2
  /home/kinow/Development/python/workspace/autosubmit/autosubmit/generators/aiida.py:146: UserWarning: Found in job POST configuration file key ADDITIONAL_FILES that is not officially supported for AiiDA generator. It might result in an error.
    warnings.warn(msg)

test/integration/generators/test_aiida.py::test_aiida_generator2
  /home/kinow/Development/python/workspace/autosubmit/autosubmit/generators/aiida.py:146: UserWarning: Found in job CLEAN configuration file key ADDITIONAL_FILES that is not officially supported for AiiDA generator. It might result in an error.
    warnings.warn(msg)

test/integration/generators/test_aiida.py::test_aiida_generator2
  /home/kinow/Development/python/workspace/autosubmit/autosubmit/generators/aiida.py:146: UserWarning: Found in job TRANSFER configuration file key ADDITIONAL_FILES that is not officially supported for AiiDA generator. It might result in an error.
    warnings.warn(msg)

```

kinow commented:
Once we get the code working here, I'd probably want to help to a) run AiiDA standalone with MN5, and then convert a local experiment to AiiDA :smiley: That's the same validation process we were planning for pyflow/ecflow, although in that case we were using the climatedt workflow which is tricky to migrate. But for AiiDA I think running the default test workflow (dummy project) is enough.

Thanks!

agoscinski committed:
Apply Brunos review

VindeeR reviewed (changes requested):
The changes are approved, but I think you might've deleted the test file by mistake, could you recreate them please, sorry for the delay.

Could you address the optional dependency @kinow had requested too.

Thanks and great job :raised_hands:
